File: LeetCode/Problems/Python/47. Permutations II.py
#my first try
#Time: O(n*n!)
#Space: O(n!)
class Solution:
    def backtrack(self, nums, result, curr):
        if len(nums) == 0:
            result.add(tuple(curr))
            
        for i in range(len(nums)):
            self.backtrack(nums[:i]+nums[i+1:], result, curr + [nums[i]])
        
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return [] 
        
        result = set()
        self.backtrack(nums, result, [])
        return result

#use sorting first to handle duplicate
#Time: O(n*n!)
#Space: O(n!)
class Solution:
    def backtrack(self, nums, result, curr):
        if len(nums) == 0:
            result.append(curr)
            
        for i in range(len(nums)):
            if i == 0 or nums[i]!= nums[i-1]:
                self.backtrack(nums[:i]+nums[i+1:], result, curr + [nums[i]])
        
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return [] 
        
        result = []
        nums.sort()
        self.backtrack(nums, result, [])
        return result
    

File: LeetCode/Problems/C++/47. Permutations II.cpp
class Solution {
public:
    bool backtracking(vector<int> nums, vector<vector<int>>& ret, int first){
        if(first == nums.size()-1){
            ret.emplace_back(nums);
            return true;
        }
        
        for(int i=first;i<nums.size();i++){
            //cout<< "first: " << first <<endl;
            if(i!=first && nums[i]==nums[first]) continue;
            
            //cout<<first << " " << nums[i] << " "<< nums[first] << endl;
            swap(nums[i], nums[first]);
            backtracking(nums, ret, first+1);
            //swap(nums[i], nums[first]);
            
        }
        return true;
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        if(nums.size()==0) return {};
        vector<vector<int>> ret;
        sort(nums.begin(),nums.end());
        backtracking(nums, ret, 0);
        
        return ret;
    }
};